### PR TITLE
Integrate Lerna and split packages

### PR DESCRIPTION
**Pull request by manuel-jasso**
Integrate Lerna and split into 3 public packages:
- @locker/near-membrane (base package)
- @locker/dom-membrane (imports @locker/near-membrane)
- @locker/node-membrane (imports @locker/near-membrane)

Karma, Jest and 'examples' are working but there are a few things to discuss:
- Whether to bundle @locker/near-membrane with @locker/dom-membrane and @locker/node-membrane (currently not bundled but it used to be), and 'bundlesize' tests need to be adjusted accordingly.
- Publishing is manual from localhost, we should have a CI job (Locker is the same)

### CODE REVIEW

File: babel.config.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
```
Comment (manuel-jasso) [inline babel.config.js:1]: This is for Jest but WIP, final form TBD.

File: examples/custom-elements/index.html [hunk 2]
```diff
@@ -2,6 +2,6 @@
     <head>
     </head>
     <body>
-        <script type="module" src="./errors.js"></script>
+        <script type="module" src="./bundle.js"></script>
```
Comment (manuel-jasso) [inline examples/custom-elements/index.html:5]: I moved all examples inside their own folder and each one now loads a  'bundle.js' file that is built.

File: examples/custom-elements/index.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import createSecureEnvironment from '../lib/browser-realm.js';
+import createSecureEnvironment from '@locker/dom-membrane';
```
Comment (manuel-jasso) [inline examples/custom-elements/index.js:1]: All examples have this same change

File: examples/package.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "examples",
+    "version": "0.4.0-test.0",
+    "private": true,
```
Comment (manuel-jasso) [inline examples/package.json:4]: 'examples' is a private package in the monorepo, it consumes @locker/dom-membrane

File: examples/rollup.config.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        input: `${dir}/index.js`,
+        output: {
+            file: `${dir}/bundle.js`,
+            format: 'cjs'
+        },
+        plugins: [nodeResolve()]
+    }
+};
+
+export default [
```
Comment (manuel-jasso) [inline examples/rollup.config.js:14]: We build a bundle.js for each example

File: jest.config.js [hunk 6]
```diff
@@ -1,13 +1,13 @@
 module.exports = {
     "roots": [
-      "<rootDir>/src"
+      "<rootDir>/packages/node-membrane/src"
```
Comment (manuel-jasso) [inline jest.config.js:3]: Jest config not working yet

File: karma.config.js [hunk 7]
```diff
@@ -22,6 +24,9 @@ module.exports = function(config) {
         format: 'es',
         sourcemap: 'inline',
       },
+      plugins: [
+        resolve(),
```
Comment (manuel-jasso) [inline karma.config.js:28]: We now have to bundle karma tests with node-resolve (like we do in Locker)

File: lerna.json [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{
+  "version": "0.4.0-test.0",
```
Comment (manuel-jasso) [inline lerna.json:2]: To test, I published this version

File: package.json [hunk 9]
```diff
@@ -1,17 +1,28 @@
 {
-  "name": "@locker/near-membrane",
-  "version": "0.4.0",
+  "name": "@salesforce/near-membrane",
+  "private": true,
```
Comment (manuel-jasso) [inline package.json:3]: The root package now has to be private under Lerna

File: package.json [hunk 10]
```diff
@@ -1,17 +1,28 @@
...
   "module": "lib/index.js",
   "types": "types/index.js",
   "scripts": {
-    "build": "yarn build:types && rollup -c && yarn build:examples",
-    "build:types": "tsc --declaration true --emitDeclarationOnly true --declarationDir types",
-    "build:examples": "rollup -i examples/web-components/platform.js -o examples/web-components/bundle.js -f cjs",
-    "start": "rollup -c --watch -w & http-server ./examples",
+    "clean": "lerna run clean --stream",
+    "build": "lerna run build --stream",
+    "start": "http-server ./examples",
```
Comment (manuel-jasso) [inline package.json:14]: Still need to work on making 'watch' work with Lerna/Rollup

File: package.json [hunk 11]
```diff
@@ -1,17 +1,28 @@
...
     "test": "jest --runInBand --no-cache --config ./jest.config.js",
     "test:bundlesize": "bundlesize",
-    "publish": "npm publish --access public",
+    "version": "yarn build",
+    "bump": "lerna version --exact --no-push --yes",
+    "bump:patch": "lerna version patch --exact --no-push --yes",
+    "bump:minor": "lerna version minor --exact --no-push --yes",
+    "bump:beta": "lerna version prerelease --preid beta --exact --no-push --yes",
+    "bump:test": "lerna version prerelease --preid test --exact --no-push --yes",
```
Comment (manuel-jasso) [inline package.json:22]: Adde various bump scripts, we can fine tune which ones to keep, but all make sense I think

File: package.json [hunk 12]
```diff
@@ -1,17 +1,28 @@
...
+    "push": "git push && git push --tags",
```
Comment (manuel-jasso) [inline package.json:23]: For now, we need a manual push from localhost until we have a CI job for publishing

File: package.json [hunk 13]
```diff
@@ -1,17 +1,28 @@
...
+    "npm:login": "npm login --registry https://registry.npmjs.org/",
+    "publish": "lerna publish from-package --ignore-scripts --yes --registry https://registry.npmjs.org/",
```
Comment (manuel-jasso) [inline package.json:25]: These scripts are to make sure we login and publish to the right registry

File: package.json [hunk 14]
```diff
@@ -49,17 +65,17 @@
   },
   "bundlesize": [
     {
-      "path": "lib/index.min.js",
+      "path": "packages/near-membrane/lib/near-membrane.min.js",
```
Comment (manuel-jasso) [inline package.json:68]: Bundlesizes need to be adjusted after we finalize bundling strategy

File: packages/dom-membrane/tsconfig.rollup.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+{
+    "extends": "../../tsconfig.json"
```
Comment (manuel-jasso) [inline packages/dom-membrane/tsconfig.rollup.json:2]: I'm still fine-tuning tsconfig files, maybe we just need two: one for types and one for rollup

**Comment by jdalton** [comment]
I think `@locker/dom-membrane` should live in `salesforce/locker/packages/@locker/dom-membrane`.

**Comment by manuel-jasso** [comment]
> I think `@locker/dom-membrane` should live in `salesforce/locker/packages/@locker/dom-membrane`.

I went back and forth on that as well... Right now our Lerna setup is fixed version (i.e. all packages have the same version), as opposed to independent. If we want to keep it like that, I think we want to have in the monorepo packages that are likely to change as a group, always having the same version bump. In this sense, I think @locker/dom-membrane is a bit more independent, and packages such as Oasis will use it without using any other Locker vNext feature.

Also, if we migrate @locker/dom-membrane outside of @salesforce/near-membrane, I think we'd also have to migrate @locker/node-membrane as well, but where?

Also, for testing @locker/near-membrane, right now we test via @locker/dom-membrane or @locker/node-membrane, so we'd have to figure how to test it instead.

We should definitely discuss this...

**Comment by jdalton** [comment]
@manuel-jasso 
> Also, for testing `@locker/near-membrane`, right now we test via `@locker/dom-membrane` or `@locker/node-membrane`, so we'd have to figure how to test it instead.

Right, so we move them out to the respective packages and leave the lower level tests in the `near-membrane` repo. We could also mock some scenarios similar to what's done in [@locker/eslint-rule-maker/test/rules](https://github.com/salesforce/locker/tree/master/packages/%40locker/eslint-rule-maker/test/rules).

File: package.json [hunk 16]
```diff
@@ -1,17 +1,28 @@
 {
-  "name": "@locker/near-membrane",
-  "version": "0.4.0",
+  "name": "@salesforce/near-membrane",
```
Comment (caridy) [inline package.json:2]: `@locker/near-membrane-multipack`

File: packages/dom-membrane/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,2 @@
+export { default } from './browser-realm';
+export * from './window';
```
Comment (caridy) [inline packages/dom-membrane/src/index.ts:2]: why is this re-exported? who is using anything other than default?

Comment (manuel-jasso) [reply packages/dom-membrane/src/index.ts]: Actually it was unnecessary, removed it, not sure why I had it 🤷🏻‍♂️

**Review by caridy — APPROVED** [review]

**Pull request merged**